### PR TITLE
feat(server): reference field

### PR DESCRIPTION
**Pull request by nourbalaha**
# Overview
- Reference field implementation has been added in this PR.
- Add reference field resolver.
- Update items query to get items by model instead of schema.
- Add `isItemReferenced` query.

### CODE REVIEW

File: server/internal/usecase/interfaces/schema.go [hunk 1]
```diff
@@ -12,6 +12,7 @@ import (
 )
 
 type CreateFieldParam struct {
+	ModelId      id.ModelID
 	SchemaId     id.SchemaID
```
Comment (rot1024) [inline server/internal/usecase/interfaces/schema.go:16]:
```suggestion
	ModelID      id.ModelID
	SchemaID     id.SchemaID
```

File: server/internal/usecase/interfaces/schema.go [hunk 2]
```diff
@@ -25,6 +26,7 @@ type CreateFieldParam struct {
 }
 
 type UpdateFieldParam struct {
+	ModelId      id.ModelID
 	SchemaId     id.SchemaID
 	FieldId      id.FieldID
```
Comment (rot1024) [inline server/internal/usecase/interfaces/schema.go:31]:
```suggestion
	ModelID      id.ModelID
	SchemaID     id.SchemaID
	FieldID      id.FieldID
```

File: server/internal/usecase/interactor/schema.go [hunk 3]
```diff
@@ -62,40 +63,245 @@ func (i Schema) CreateField(ctx context.Context, param interfaces.CreateFieldPar
...
 			return nil, err
 		}
 
-		return f, nil
+		return f1, nil
+	})
+}
+
+func (i Schema) updateCorrespondingField(ctx context.Context, f1 *schema.Field, param interfaces.UpdateFieldParam, operator *usecase.Operator) error {
+	return Run0(ctx, operator, i.repos, Usecase().Transaction(), func(ctx context.Context) error {
```
Comment (rot1024) [inline server/internal/usecase/interactor/schema.go:191]:
```suggestion
```

you should not do double transactions

File: server/internal/usecase/interactor/schema.go [hunk 4]
```diff
@@ -62,40 +63,245 @@ func (i Schema) CreateField(ctx context.Context, param interfaces.CreateFieldPar
...
+		if err := i.repos.Schema.Save(ctx, s1); err != nil {
 			return nil, err
 		}
 
-		return f, nil
+		return f1, nil
+	})
+}
+
+func (i Schema) updateCorrespondingField(ctx context.Context, f1 *schema.Field, param interfaces.UpdateFieldParam, operator *usecase.Operator) error {
```
Comment (rot1024) [inline server/internal/usecase/interactor/schema.go:190]: This func is actually domain logic, because there is few logics to load and save something with repos in the function.

```go
// pkg/schema/corresponding_field.go

type CorrespondingFieldTuple struct {
	Schema1 *Schema
	Field1  *Field
	Schema2 *Schema
	Field2  *Field
}

func GetOrCreateCorrespondingFields(s1, s2 *Schema, fid FieldID) (CorrespondingFieldTuple, error) {
	// ...
}
```

**Review by rot1024 — CHANGES_REQUESTED** [review]

File: server/internal/usecase/interfaces/schema.go (hunk 2 above)
Comment (nourbalaha) [reply server/internal/usecase/interfaces/schema.go:31]: Done!

File: server/internal/usecase/interactor/schema.go (hunk 3 above)
Comment (nourbalaha) [reply server/internal/usecase/interactor/schema.go:191]: Done!

File: server/pkg/schema/corresponding_field.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+)
+
+type CorrespondingFieldTuple struct {
+	Schema1 *Schema
+	Field1  *Field
+	Schema2 *Schema
+	Field2  *Field
+}
+
+func GetCorrespondingFields(s1, s2 *Schema, mid id.ModelID, f1 *Field, fr *FieldReference) (*CorrespondingFieldTuple, error) {
```
Comment (rot1024) [inline server/pkg/schema/corresponding_field.go:17]: add test. these logic are not tested so it should have some bugs.

File: server/pkg/schema/corresponding_field.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+	cf1 := fr.CorrespondingField()
+	// check if reference direction is two way
+	if cf1 != nil {
```
Comment (rot1024) [inline server/pkg/schema/corresponding_field.go:20]:
```suggestion
	// check if reference direction is two way
	if cf1 := fr.CorrespondingField(); cf1 != nil {
```

**Review by rot1024 — CHANGES_REQUESTED** [review]

File: server/pkg/schema/corresponding_field.go (hunk 6 above)
Comment (nourbalaha) [reply server/pkg/schema/corresponding_field.go:20]: Done!

File: server/internal/usecase/interactor/schema.go [hunk 7]
```diff
@@ -62,13 +62,88 @@ func (i Schema) CreateField(ctx context.Context, param interfaces.CreateFieldPar
...
 			return nil, err
 		}
 
-		return f, nil
+		return f1, nil
+	})
+}
+
+func (i Schema) createCorrespondingField(ctx context.Context, fr *schema.FieldReference, f1 *schema.Field, mId1 id.ModelID, operator *usecase.Operator) (*schema.Field, error) {
+	return Run1(ctx, operator, i.repos, Usecase().Transaction(), func(ctx context.Context) (*schema.Field, error) {
```
[the remark this answers is not in the document]
Comment (nourbalaha) [reply server/internal/usecase/interactor/schema.go]: Done!

File: server/pkg/schema/corresponding_field_test.go [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+	// check that fields are not nil
+	assert.NotNil(t, fields.Schema1)
+	assert.NotNil(t, fields.Schema2)
+	assert.NotNil(t, fields.Field1)
+	assert.NotNil(t, fields.Field2)
+
+	// check that model ids are correct
+	assert.Equal(t, fields.Field2.typeProperty.reference.modelId, mid1)
+	assert.Equal(t, fields.Field1.typeProperty.reference.modelId, mid2)
```
Comment (rot1024) [inline server/pkg/schema/corresponding_field_test.go:39]: You may want to minimize the use of private fields in your test code that are not directly related to the implementation you are trying to test.You never know when a private field will change and its specification is implicit. Using public methods makes it clear that the intent is not to test them.

```suggestion
	rf, _ := FieldReferenceFromTypeProperty(fields.Field2.TypeProperty())
	assert.Equal(t, rf.modelId, mid1)
	rf, _ := FieldReferenceFromTypeProperty(fields.Field1.TypeProperty())
	assert.Equal(t, rf.modelId, mid2)
```

**Review by rot1024 — APPROVED** [review]

File: server/schemas/field.graphql [hunk 9]
```diff
@@ -121,7 +121,9 @@ type SchemaFieldInteger {
 
 type SchemaFieldReference {
   modelId: ID!
+  correspondingSchemaId: ID
   correspondingFieldId: ID
+  correspondingField: SchemaField
```
Comment (rot1024) [inline server/schemas/field.graphql:126]: Also you can add some fileds (with resolver) like

```suggestion
  model: Model
  correspondingSchema: Schema
  correspondingField: SchemaField
```

File: server/internal/adapter/gql/loader_item.go [hunk 10]
```diff
@@ -106,22 +108,30 @@ func (c *ItemLoader) FindVersionedItems(ctx context.Context, itemID gqlmodel.ID)
...
 	}
 
-	s, err := c.schemaUsecase.FindByID(ctx, sid, op)
+	s, err := c.schemaUsecase.FindByID(ctx, m.Schema(), op)
 	if err != nil {
 		if errors.Is(err, rerror.ErrNotFound) {
 			return nil, nil
 		}
 		return nil, err
 	}
```
Comment (rot1024) [inline server/internal/adapter/gql/loader_item.go:132]: they are not needed I think

Comment (nourbalaha) [reply server/internal/adapter/gql/loader_item.go:132]: I think that schema is needed for this line of code:
`itm := gqlmodel.ToItem(i, s)`

File: server/schemas/field.graphql (hunk 9 above)
Comment (nourbalaha) [reply server/schemas/field.graphql:126]: Done!
